### DISCUSSION TITLE
Is this correct?

### DISCUSSION THREAD

**Discussion by torsteinelv** [syssi/esphome-jk-bms#195] [General]
Hello,

Just a question to try to understand the readings from this library.
Not sure why balacing is Off when its activated? also it says max cell voltage 7 and min cell voltage 1. Is this 7 cell that is over and 1 is under? Or do i have some kind of an error here?

![Screenshot_20221017-094549_Home Assistant](https://user-images.githubusercontent.com/18403384/196119955-09aaab6c-a4f6-4a42-a197-71fd40e57412.jpg)

**Comment by syssi** [comment]
The "min voltage cell" indicates the cell number with the lowest cell voltage (at the moment). The max voltage cell is the cell with the highest voltage. In case of balancing the balancer will move some energy from the `max voltage cell` to the `min voltage cell`.

The setting `cell voltage overvoltage` and `cell voltage undervoltage` are settings to protect your cells from over-/under-charging.

If you turn on the `balancer` switch it doesn't mean the BMS will start balancing immediately. There are some constraints that must be met. For example:

1. The `delta cell voltage` (max cell voltage - min cell voltage) must be greater than `balance trigger voltage` (0.01V)
2. The cell voltage of all cells must be greater than `balacing starting voltage` (3.3V)

It looks like the first constraint isn't met: `0,001V < 0,01V` -> The balancing indicator (binary sensor) is off.

**Reply by torsteinelv** [reply]
aha. looks like mye battery keeping it self balanced without help from the jk bms? When i use the app I can see a red cell and a blue. seems like the bms working with balacing. But I also see that bms balance current is 0. So I guess its just to little for it to care about.

Thanks for your reply. 

I have some wierd abnormal sensor in my jk bms. Is it possible to see this in error sensor to figure out why this happens?

**Reply by syssi** [reply]
The implementation of the error/protection bitmask is incomplete: https://github.com/syssi/esphome-jk-bms/blob/main/components/jk_bms_ble/jk_bms_ble.cpp#L25-L42

You can use the 

```
switch:
  - platform: ble_client
    ble_client_id: client0
    name: "${name} enable bluetooth connection"
```

to disconnect the BLE connection to the BMS. So you are able to use the Android App. If you see an unnamed error (f.e. `Error 0x00 0x10`) via ESPHome please try to identify the related error message of the Android App. I would like to complete the error bitmask.

**Reply by torsteinelv** [reply]
Sure no problem. But I can only see this in the console? I can just log this to an output file and it will not be long before we have abnormal release current sensor.

**Reply by pato-loco** [reply]
> The implementation of the error/protection bitmask is incomplete: https://github.com/syssi/esphome-jk-bms/blob/main/components/jk_bms_ble/jk_bms_ble.cpp#L25-L42
> 
> You can use the
> 
> ```
> switch:
>   - platform: ble_client
>     ble_client_id: client0
>     name: "${name} enable bluetooth connection"
> ```
> 
> to disconnect the BLE connection to the BMS. So you are able to use the Android App. If you see an unnamed error (f.e. `Error 0x00 0x10`) via ESPHome please try to identify the related error message of the Android App. I would like to complete the error bitmask.

Error 0x40 0x00 = charge overcurrent protection

**Reply by syssi** [reply]
Thanks for you feedback! I've pushed a change: https://github.com/syssi/esphome-jk-bms/commit/50a879522bc00ccbfb0d60fc2de2b55bc4d49521

If you are using the `main` branch you could rebuild & flash your ESP to benefit from this change.

**Reply by pato-loco** [reply]
> Thanks for you feedback! I've pushed a change: [50a8795](https://github.com/syssi/esphome-jk-bms/commit/50a879522bc00ccbfb0d60fc2de2b55bc4d49521)
> 
> If you are using the `main` branch you could rebuild & flash your ESP to benefit from this change.

in Zeile 1113 (jk_bms_ble.cpp) ist nun ein ";" zu viel

**Reply by syssi** [reply]
Wie kommst du darauf?

**Reply by pato-loco** [reply]
> Wie kommst du darauf?

Compiling .pioenvs\pv01bat01mon\src\main.o
src/esphome/components/jk_bms_ble/jk_bms_ble.cpp: In member function 'bool esphome::jk_bms_ble::JkBmsBle::write_register(uint8_t, uint32_t, uint8_t)':
src/esphome/components/jk_bms_ble/jk_bms_ble.cpp:1211:115: warning: suggest braces around empty body in an 'if' statement [-Wempty-body]
     ESP_LOGW(TAG, "[%s] esp_ble_gattc_write_char failed, status=%d", this->parent_->address_str().c_str(), status);
                                                                                                                   ^

**Reply by syssi** [reply]
Ist nur eine Warnung. Ich habe zusaetzliche Klammern eingeführt in der Hoffnung die Warnung zu lösen.

**Comment by torsteinelv** [comment]
![Screenshot_20221017-163003_Home Assistant](https://user-images.githubusercontent.com/18403384/196204427-82c5019d-f3b5-4bbc-ae20-a39bba202beb.jpg)
Is this some errors?

**Reply by syssi** [reply]
Do you have access to a linux host? So we could install `esphome` here to consume the logs more easily. Your screenshot shows some raw traffic of the log level `DEBUG`.

**Reply by torsteinelv** [reply]
I will install a linux server later today.

**Reply by syssi** [reply]
Please follow this screen cast to install esphome:

![Screen recording](https://raw.githubusercontent.com/syssi/esphome-seplos-bms/main/install.gif)

**Comment by torsteinelv** [comment]
OK. Im ready now to collect the information you need if you tell me how to do it :)

**Reply by syssi** [reply]
Are you sure something doesn't work as expected? I assume you are confused because of the log level `DEBUG` messages. You could decrease the log level to get rid of these messages by:

```
logger:
  level: INFO
```

Does all sensor work as expected? Does they show meaningful measurements?

**Reply by torsteinelv** [reply]
Yes. You are right. Etabled info now and logging to file. Will report when I have translated the errors.

**Reply by syssi** [reply]
Thanks for your support!
